tests/cpp/test_matmul_aten_evaluation.cpp

```diff
@@ -393,4 +408,79 @@ TEST_F(MatmulATenEvaluationTest, LinearWithBias) {
 
   EXPECT_TRUE(at::allclose(out[0], out_ref));
 }
+
+TEST_P(ATenNodesParametrizedTest, MatmulNodeConcrete) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const auto& [a_shape, b_shape] = GetParam();
+
+  auto tv0 = makeConcreteTensor(a_shape, DataType::Half);
+  auto tv1 = makeConcreteTensor(b_shape, DataType::Half);
+  auto tv2 = eagerMatmul(tv0, tv1);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv2);
+
+  at::Tensor t0 = at::randn(a_shape, at::kHalf).cuda();
+  at::Tensor t1 = at::randn(b_shape, at::kHalf).cuda();
+  at::Tensor out_ref = at::matmul(t0, t1);
+
+  FusionExecutor fe;
+  fusion->aliasOutputToInput(
+      fusion->outputs()[0], /*input=*/nullptr, AllocationType::Evaluate);
+  fe.compileFusion(fusion.get(), {t0, t1});
+  auto out = fe.runFusion({t0, t1});
+
+  // Verify that fusion compilation was skipped.
+  EXPECT_FALSE(fe.hasCompiledKernel());
+
+  EXPECT_TRUE(at::allclose(out[0], out_ref));
+}
+
+TEST_P(ATenNodesParametrizedTest, MatmulNodeSymbolic) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  const auto& [a_shape, b_shape] = GetParam();
+
+  auto tv0 = makeSymbolicTensor(a_shape.size(), DataType::Half);
+  auto tv1 = makeSymbolicTensor(b_shape.size(), DataType::Half);
+  auto tv2 = eagerMatmul(tv0, tv1);
+
+  fusion->addInput(tv0);
+  fusion->addInput(tv1);
+  fusion->addOutput(tv2);
+
+  at::Tensor t0 = at::randn(a_shape, at::kHalf).cuda();
+  at::Tensor t1 = at::randn(b_shape, at::kHalf).cuda();
+  at::Tensor out_ref = at::matmul(t0, t1);
+
+  FusionExecutor fe;
+  fusion->aliasOutputToInput(
+      fusion->outputs()[0], /*input=*/nullptr, AllocationType::Evaluate);
+  fe.compileFusion(fusion.get(), {t0, t1});
+  auto out = fe.runFusion({t0, t1});
+
+  // Verify that fusion compilation was skipped.
+  EXPECT_FALSE(fe.hasCompiledKernel());
+
+  EXPECT_TRUE(at::allclose(out[0], out_ref));
+}
+
+constexpr int64_t b = 128, m = 64, k = 32, n = 16;
+
+INSTANTIATE_TEST_SUITE_P(
+    ,
+    ATenNodesParametrizedTest,
+    testing::Values(
+        std::make_tuple(Sizes({k}), Sizes({k})),
+        std::make_tuple(Sizes({m, k}), Sizes({k, n})),
+        std::make_tuple(Sizes({k}), Sizes({k, n})),
+        std::make_tuple(Sizes({m, k}), Sizes({k})),
+        std::make_tuple(Sizes({k}), Sizes({b, k, n})),
+        std::make_tuple(Sizes({b, m, k}), Sizes({k})),
+        std::make_tuple(Sizes({b, 1, m, k}), Sizes({b, k, n}))));
```

Comment:
Is it true that we can accept any combination of A and B where A is one of `{k}, {m, k}, {b, m, k}, {b, 1, m, k}` and B is one of `{k}, {k, n}, {b, k, n}`? If so maybe we could parametrize each of those separately and hit all combos.

Reply:
Out of curiosity. What would happen if k/m/n happen to be 1?

Reply:
> Out of curiosity. What would happen if k/m/n happen to be 1?

Do you mean if the output shape is different? -- It will be the same as any other case.

Or how we intend to handle those cases?
@jacobhinkle pointed out we could special-case these cases though to increase opportunity for fusion. (For eg: For `[M, 1] x [1, N]`, we can simply return the outer product without creating the `MatmulOp` node)

Reply:
> For `[M, 1] x [1, N]`

This is the `K=1` case but I think Jie is asking what about M and N, which we aren't testing here. I would also add that we're not really testing this case properly either since we are creating the input tensors with `makeSymbolicTensor(a_shape.size())` which will make all dimensions `IterType::Iteration`. I suggested a code change that I think will address that, then we can add shape combos that have 1s in each position.

Reply:
I added cases for M=1/N=1. At present, they will behave the same way as when M/N > 1.

Reply:
>  I would also add that we're not really testing this case properly either since we are creating the input tensors with `makeSymbolicTensor(a_shape.size())` which will make all dimensions `IterType::Iteration`. 

`makeSymbolicTensor(a_shape)` -> Does this mark dimensions as Broadcast if the extent is 1?